### PR TITLE
Fix bitsandbytes layer conversion under `init_module` context manager

### LINKED ISSUE #18913: Bitsandbytes doesn't convert layers under `init_module` context
### Bug description

The strategy calls the `tensor_init_context()` instead of the `module_init_context()`, causing `BitsandbytesPrecision.init_module_context()` not to be called.

### What version are you seeing the problem on?

v2.1

### How to reproduce the bug

```python
fabric = Fabric(accelerator="cpu", plugins=BitsandbytesPrecision("int8-training"))
with fabric.init_module():
    model = torch.nn.Linear(2, 2)

print(type(model))  # returns nn.Linear()
```

From Pietro:
```py
from lightning.fabric import Fabric
from transformers import AutoModelForSequenceClassification
from lightning.fabric.plugins.precision import BitsandbytesPrecision

fabric = Fabric(accelerator="gpu", plugins=BitsandbytesPrecision(mode="nf4"))

with fabric.init_module():
	mymodel = AutoModelForSequenceClassification.from_pretrained("bert-base-uncased")

print(fabric._precision.precision)
#> '32-true'

print(mymodel.bert.encoder.layer[4].output.dense)
#> Linear(in_features=3072, out_features=768, bias=True)

setup_mymodel = fabric.setup(mymodel)

print(setup_mymodel.bert.encoder.layer[4].output.dense)
#> _NF4Linear(in_features=3072, out_features=768, bias=True)
```

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0):
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 2.0):
#- Python version (e.g., 3.9):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):
```

</details>

### More info

Reported by @pietrolesci 

cc @carmocca

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #18913

This test fails on master, but didn't run in our CI because GPU CI does not install bitsandbytes:
https://github.com/Lightning-AI/lightning/blob/7a5b7f5561b34f13da1e0aa26af18b94b62494cc/tests/tests_fabric/plugins/precision/test_bitsandbytes.py#L102

This PR adds the missing installation. Now the test passes:

https://dev.azure.com/Lightning-AI/72ab7ed8-b00f-4b6e-b131-3388f7ffafa7/_build/results?buildId=180654&view=logs&j=3f274fac-2e11-54ca-487e-194c91f3ae9f

```
plugins/precision/test_bitsandbytes.py::test_bitsandbytes_layers[args0-expected0] PASSED [ 37%]
plugins/precision/test_bitsandbytes.py::test_bitsandbytes_layers[args1-expected1] PASSED [ 38%]
```

----
:books: Documentation preview :books:: https://pytorch-lightning--18914.org.readthedocs.build/en/18914/

cc @carmocca @borda @justusschock @awaelchli

### CODE REVIEW

**Review by carmocca — APPROVED** [review]

**Comment by awaelchli** [comment]
What would be the best way to install bitsandbytes in the GPU CI @carmocca @Borda ? Should I put it into the workflow or have it in the docker image?

**Comment by carmocca** [comment]
I think the docker image. We could put it in `extra.txt` but then users might have issues if it fails to install in certain platform (windows? macos?)

File: requirements/pytorch/extra.txt [hunk 1]
```diff
@@ -8,3 +8,4 @@ hydra-core >=1.0.5, <1.4.0
 jsonargparse[signatures] >=4.26.1, <4.27.0
 rich >=12.3.0, <13.6.0
 tensorboardX >=2.2, <2.7.0  # min version is set by torch.onnx missing attribute
+bitsandbytes <=0.41.1
```
Comment (Borda) [inline requirements/pytorch/extra.txt:11]:
```suggestion
bitsandbytes <0.42.0
```
Could it be less strict?

**Comment by awaelchli** [comment]
The docker build passed, I pushed the image, re-triggered the GPU CI and the test passes now!

**Review by Borda — APPROVED** [review]

File: requirements/pytorch/extra.txt (hunk 1 above)
Comment (awaelchli) [reply requirements/pytorch/extra.txt:11]: I think it's better to let dependabot bump it when a new version comes out.

**Pull request merged**